### PR TITLE
SafeCrLf and AutoCrLF improvements to make libgit2 behave like vanilla git

### PR DESCRIPTION
**Pull request by csware**
If SafeCrLf=true and AutoCrLf=true hold libgit2 right now accepts a LF only file, vanilla git fails with "fatal: LF would be replaced by CRLF in [FILENAME]" (on Windows)

Also handling of malformed files (containing LF, CR and CRLF) was improved. And support for "\* text eol=crlf" was added for LF-only platforms.

See change log file for all changes.

### CODE REVIEW

File: tests/filter/crlf.c [hunk 1]
```diff
@@ -103,12 +103,12 @@ void test_filter_crlf__with_safecrlf(void)
...
 	cl_assert_equal_i(giterr_last()->klass, GITERR_FILTER);
 
-	/* Normalized \n is reversible, so does not fail with safecrlf */
+	/* Normalized \n is reversible, however, vanilla git does fail here -> see tests/index/crlf.c */
 	in.ptr = "Normal\nLF\nonly\nline-endings.\n";
 	in.size = strlen(in.ptr);
 
-	cl_git_pass(git_filter_list_apply_to_data(&out, fl, &in));
-	cl_assert_equal_s(in.ptr, out.ptr);
+	cl_git_fail(git_filter_list_apply_to_data(&out, fl, &in));
```
Comment (csware) [inline tests/filter/crlf.c]: I'm unsure about this change as the original comment said "Normalized \n is reversible, so does not fail with safecrlf", however, vanilla git refuses to handle this kind of problem...

**Comment by YueLinHo** [comment]
About the tests, I trace to 16798d08cf3a1757deb9f0363b35fbf775cfc3fb.
About source code, trace to c094197bf92736bb1c40cf1ca87bda970ab7f999
@arrbee could you please have a look at this?
This PR might be a regression of them?

**Comment by YueLinHo** [comment]
One more question?
Does the test test_filter_crlf__with_safecrlf() run under AutoCrlf = true?

If not or unknown, I suppose the test is still broken.
The reason is SafeCrlf should be considered with AutoCrlf together.
That says SafeCrlf only effective under AutoCrlf = true.
And the tests should be modified by adding AutoCrlf part.

**Comment by YueLinHo** [comment]
cross ref. https://github.com/git/git/blob/master/convert.c#L176

**Comment by csware** [comment]
@ethomson Could you please have a look at this?

File: src/crlf.c [hunk 2]
```diff
@@ -138,8 +138,8 @@ static int crlf_apply_to_odb(
 		if (git_buf_text_gather_stats(&stats, from, false))
 			return GIT_PASSTHROUGH;
 
-		/* If there are no CR characters to filter out, then just pass */
-		if (!stats.cr)
+		/* If there are no CR characters to filter out and CrLf is not set to "true", then just pass */
+		if (!stats.cr && ca->auto_crlf != GIT_AUTO_CRLF_TRUE)
```
Comment (csware) [inline src/crlf.c]: This patch perfectly works on *nix and the clar tests however not on Win32, as auto_crlf is not properly initialized based on core.autocrlf :(

**Comment by csware** [comment]
The current PR makes libgit2 behave like vanilla giut described in the table above by @YueLinHo.

There's only one thing I'm a bit concerned about is, that libgit2 with SafeCrLf=true is much more picky about files like "abc\rdef\nfff" which fail in libgit2 but work with vanilla git - dunno if that is an error in git.

File: tests/checkout/crlf.c [hunk 3]
```diff
@@ -313,6 +327,78 @@ void test_checkout_crlf__autocrlf_false_text_auto_attr(void)
...
+	git_checkout_options opts = GIT_CHECKOUT_OPTIONS_INIT;
+	opts.checkout_strategy = GIT_CHECKOUT_SAFE_CREATE;
+
+	cl_git_mkfile("./crlf/.gitattributes", "* crlf\n");
+
+	cl_repo_set_bool(g_repo, "core.autocrlf", false);
+
+	git_checkout_head(g_repo, &opts);
+
+	check_file_contents("./crlf/all-lf", ALL_LF_TEXT_AS_CRLF);
```
Comment (csware) [inline tests/checkout/crlf.c]: This test fails and I don't know why ATM

File: src/buf_text.c [hunk 4]
```diff
@@ -125,16 +125,14 @@ int git_buf_text_lf_to_crlf(git_buf *tgt, const git_buf *src)
 		size_t copylen = next - scan;
 		size_t needsize = tgt->size + copylen + 2 + 1;
 
-		/* if we find mixed line endings, bail */
```
Comment (csware) [inline src/buf_text.c]: Handling is now done in crlf.c

**Comment by csware** [comment]
This PR makes libgit2 behave like vanilla git regarding AutoCRLF aspects. The table of @YueLinHo is now fully supported (w/o the warning as the warning API is missing) and lots of other issues are fixed now.

Finally this is finished and double checked. All my changes are backed by tests (I added lots of them, based on the output of vanilla git, see https://download.tortoisegit.org/tmp/crlf-multitest.tar.bz2 for a test script which reports all CRLF fatal errors as well as the OIDs after adding files).

For all changes please see the changelog file.

@carlosmn @ethomson @arrbee  I think this is ready for review now.

**Comment by csware** [comment]
Rebased...

**Comment by nulltoken** [comment]
@csware A quick question: Would that allow us to unskip some of LibGit2Sharp tests (see https://github.com/libgit2/libgit2sharp/blob/vNext/LibGit2Sharp.Tests/BlobFixture.cs#L225) which retreive the content of a blob from the odb?

**Comment by csware** [comment]
@nulltoken I'd say so. I added tests for all combinations and verified the behaviors on Windows and *nix.

**Comment by nulltoken** [comment]
@carlosmn @Therzok Could you please give a try to https://github.com/libgit2/libgit2sharp/tree/ntk/nix_autocrlf on a non Windows platform?

This would require you to build a custom libgit2 bin based on this PR. I haven't been able to update the submodule as the change isn't known from libgit2 yet.

**Comment by Therzok** [comment]
@nulltoken I'm getting a weird build error on OS X.

```
[  0%] Building C object CMakeFiles/git2.dir/src/annotated_commit.c.o
In file included from /Users/therzok/libgit2sharp/libgit2/src/annotated_commit.c:8:
In file included from /Users/therzok/libgit2sharp/libgit2/src/common.h:68:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/usr/include/regex.h:103:1: error:
      invalid storage class specifier in function declarator
typedef __darwin_off_t regoff_t;
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/usr/include/regex.h:103:24: error:
      parameter named 'regoff_t' is missing
typedef __darwin_off_t regoff_t;
```

**Comment by Therzok** [comment]
@nulltoken you could modify the script to add another remote to the libgit2 submodule.

```
cd libgit2
git fetch origin refs/pull/2798/head:refs/remotes/pr/2798
git checkout <hash_here>
```

**Comment by nulltoken** [comment]
@Therzok @csware I've opened https://github.com/libgit2/libgit2sharp/pull/961 to test this out.

 - Windows passes
 - Mac OS X complains and indeed refuses to build (which is strange as this PR builds fine. I'll take a deeper look at the build script tomorrow if noone beats me to it first)
 - BlobFixture.cs tests still fail on Linux. So either the tests expectations are dead wrong (which is quite possible) or those use cases aren't covered by this PR.

**Comment by csware** [comment]
@nulltoken check out the build results on *nix: The error there is that a .dll is not found in repository_open...

**Comment by nulltoken** [comment]
@csware the Dll isn't found on OSX because the native build didn't happen, as far as I understand.
However, on Linux, dll is built, tests are run and pass except the ones below.


```
/home/travis/build/libgit2/libgit2sharp/CI/build.msbuild (Deploy) ->
(Test target) ->
	/home/travis/build/libgit2/libgit2sharp/CI/build.msbuild: error : LibGit2Sharp.Tests.BlobFixture.CanGetBlobAsFilteredText(autocrlf: "true", expectedText: "hey there
"): Assert.Equal() Failure
Position: First difference is at position 9
Expected: hey there
          
Actual:   hey there
          
	/home/travis/build/libgit2/libgit2sharp/CI/build.msbuild: error :   at LibGit2Sharp.Tests.BlobFixture.CanGetBlobAsFilteredText (System.String autocrlf, System.String expectedText) [0x0002d] in /home/travis/build/libgit2/libgit2sharp/LibGit2Sharp.Tests/BlobFixture.cs:39 
  at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0 
	/home/travis/build/libgit2/libgit2sharp/CI/build.msbuild: error : LibGit2Sharp.Tests.BlobFixture.CanReadBlobFilteredStream(autocrlf: "true", expectedContent: "hey there
"): Assert.Equal() Failure
Expected: 11
Actual:   10
	/home/travis/build/libgit2/libgit2sharp/CI/build.msbuild: error :   at LibGit2Sharp.Tests.BlobFixture.CanReadBlobFilteredStream (System.String autocrlf, System.String expectedContent) [0x0003e] in /home/travis/build/libgit2/libgit2sharp/LibGit2Sharp.Tests/BlobFixture.cs:140 
  at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0 
```

**Comment by csware** [comment]
@nulltoken Can you directly point me to the failed test? What's it testing?

**Comment by nulltoken** [comment]
>@nulltoken Can you directly point me to the failed test? What's it testing?

Below are the tests that exercise these autocrlf settings:

- https://github.com/libgit2/libgit2sharp/blob/vNext/LibGit2Sharp.Tests/BlobFixture.cs#L26-L45
- https://github.com/libgit2/libgit2sharp/blob/vNext/LibGit2Sharp.Tests/BlobFixture.cs#L128-L153

They basically wraps a call to `git_blob_filtered_content_stream()`

**Comment by csware** [comment]
@nulltoken The test in BlobFixture looks borked as on *nix no CR is added if AutoCrLf is set. But your test seems to test for that.

**Comment by nulltoken** [comment]
@csware Wow. Thanks for this. Could you please point me to some official documentation or something (regarding CRLF handling on *nix)? I keep on finding contradictory answers on this topic (eg. https://groups.google.com/forum/#!topic/git-users/ck_pKFGCYKc). Even the git-config doc doesn't seem to refer to some special handling... My google-fu may be quite lacking, though...

**Comment by csware** [comment]
@nulltoken Sorry, I performed lots of tests in order to see how git handles different cases on different platforms. AutoCrLf removes CrLf on all systems, but only adds it on CrLf systems. if you want to have CrLf on *nix systems you have to use .gitattributes and set e.g. "*.txt text eol=crlf"

**Comment by nulltoken** [comment]
@csware Thanks for the feedback. I'll update this PR with test coverage exercising `eol` and will keep you updated.

**Comment by csware** [comment]
@nulltoken: See https://github.com/git/git/blob/master/convert.c#L259 and https://github.com/git/git/blob/master/convert.c#L117. On Lf-only platforms EOL_NATIVE=EOL_LF is returned and on checkout no Cr is added.

**Comment by peff** [comment]
@csware I think using the new text/eol gitattributes that we will never write CRLF on a Unix system, but I think `core.autocrlf` will. For example:

```
$ git init
$ git config core.autocrlf true
$ printf 'foo\r\n' >file
$ git add . && git commit -m foo
$ git cat-file blob HEAD:file | xxd
0000000: 666f 6f0a                                foo.
```

OK, so git stripped out CR going into the repository. So far so good.  Now let's have it write out the file:

```
$ rm file
$ git reset --hard
$ xxd <file
0000000: 666f 6f0d 0a                             foo..
```

It came back with a CRLF in the working tree. I think this funkiness was one of the reasons that the text/eol attributes were invented. They can ride along with the repository, because they do the right thing everywhere (whereas you would not want to force autocrlf on Unix people, because it does the wrong thing there).

**Comment by csware** [comment]
@peff: Right, I'm a bit puzzled right now and have to recheck my old testings in a quiet moment...

**Comment by peff** [comment]
@csware You would not be the first person puzzled by git's CRLF machinery. :trollface:

**Comment by csware** [comment]
@peff @nulltoken Thanks for your help. Missing tests added, tests tripple-checked and code updated.

If you want to do your own tests: https://download.tortoisegit.org/tmp/crlf-multitest.tar.bz2

**Comment by nulltoken** [comment]
Just retested through https://github.com/libgit2/libgit2sharp/pull/961

Ok fro the CRLF perspective

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,23 @@ v0.22 + 1
 * Updated binary identification in CRLF filtering to avoid false positives in
   UTF-8 files.
 
+* LF -> CRLF filter now correctly honors SafeCrLf=true errors
+  * LF only files were accepted with AutoCrLf=true on CRLF platforms
```
Comment (ethomson) [inline CHANGELOG.md]: Please let's not use camelcase here.  For describing the actual character, please use "CR" or "LF".  For describing the configuration option, please use all lowercase.  "core.autocrlf", that sort of thing.

File: src/crlf.c [hunk 6]
```diff
@@ -124,40 +124,23 @@ static int crlf_apply_to_odb(
 	const git_buf *from,
 	const git_filter_source *src)
 {
+	git_buf_text_stats stats;
+	int isBinary;
```
Comment (ethomson) [inline src/crlf.c]: `is_binary` here, please, we don't use camelcase.

File: src/crlf.c [hunk 7]
```diff
@@ -174,11 +157,45 @@ static int crlf_apply_to_odb(
 			if (has_cr_in_index(src))
 				return GIT_PASSTHROUGH;
 		}
+	}
 
-		if (!stats.cr)
-			return GIT_PASSTHROUGH;
+	/* If safecrlf is enabled, sanity-check the result. */
+	if (ca->crlf_action == GIT_CRLF_INPUT || (ca->auto_crlf == GIT_AUTO_CRLF_INPUT && (ca->crlf_action == GIT_CRLF_GUESS || ca->crlf_action == GIT_CRLF_AUTO || (ca->crlf_action == GIT_CRLF_TEXT && ca->eol==GIT_EOL_UNSET)))) {
```
Comment (ethomson) [inline src/crlf.c]: please wrap this at 80 characters

File: src/crlf.c [hunk 8]
```diff
@@ -174,11 +157,45 @@ static int crlf_apply_to_odb(
...
+					git_filter_source_path(src));
+				return -1;
+			case GIT_SAFE_CRLF_WARN:
+				/* TODO: issue warning when warning API is available */;
+				break;
+			default:
+				break;
+			}
+		}
+	} else if (ca->crlf_action == GIT_CRLF_CRLF || (ca->auto_crlf == GIT_AUTO_CRLF_TRUE && (ca->crlf_action == GIT_CRLF_GUESS || ca->crlf_action == GIT_CRLF_AUTO || (ca->crlf_action == GIT_CRLF_TEXT && ca->eol == GIT_EOL_UNSET)))) {
```
Comment (ethomson) [inline src/crlf.c]: please wrap this at 80 characters.

File: src/crlf.c [hunk 9]
```diff
@@ -225,24 +246,49 @@ static int crlf_apply_to_workdir(
 	struct crlf_attrs *ca, git_buf *to, const git_buf *from)
 {
 	const char *workdir_ending = NULL;
+	git_buf_text_stats stats;
+	int isBinary;
```
Comment (ethomson) [inline src/crlf.c]: again, please use something like `is_binary` here, or just `binary`.  and it looks like it should be a `bool`, not an `int`.

File: src/crlf.c (hunk 6 above)
Comment (ethomson) [inline src/crlf.c]: and it looks like this should be a `bool`, not an `int`.

File: tests/index/crlf.c [hunk 10]
```diff
@@ -7,9 +7,28 @@
...
+#define FILE_OID_MORE_LF "04de00b358f13389948756732158eaaaefa1448c"
+#define FILE_OID_LF_CR "d11e7ef63ba7db1db3b1b99cdbafc57a8549f8a4"
+#define FILE_OID_LF_CR_CRLF "417786fc35b3c71aa546e3f95eb5da3c8dad8c41"
+#define FILE_OID_BINARY_LF "af6fcf6da196f615d7cda269b55b5c4ecfb4a5b3"
+#define FILE_OID_BINARY_CRLF "85340755cfe5e28c2835781978bb1cece91b3d0f"
+#define FILE_OID_BINARY_LF_CR "203555c5676d75cd80d69b50beb1f4b588c59ceb"
+#define FILE_OID_BINARY_LF_CR_CRLF "aaf083a9cb53dac3669dcfa0e48921580d629ec7"
+
+#define CHECK_CRLF_TO_LF_ERROR cl_assert(giterr_last() != NULL && giterr_last()->klass == GITERR_FILTER && strstr(giterr_last()->message, "CRLF would be replaced by LF in") != NULL); giterr_clear();
+#define CHECK_LF_TO_CRLF_ERROR cl_assert(giterr_last() != NULL && giterr_last()->klass == GITERR_FILTER && strstr(giterr_last()->message, "LF would be replaced by CRLF in") != NULL); giterr_clear();
```
Comment (ethomson) [inline tests/index/crlf.c]: can we get some line breaks here?

**Comment by csware** [comment]
@ethomson Thanks for the review. All requested changed integrated.

**Comment by csware** [comment]
Rebased.

**Comment by nulltoken** [comment]
@ethomson @carlosmn Anything you'd like to see updated or is it good to go?

**Comment by csware** [comment]
Rebased.

All failing tests are not caused by the PR, but failures on Travis ("The command "./script/install-deps-${TRAVIS_OS_NAME}.sh" failed and exited with 100 during .", "Function call failed: (git_clone(&g_repo, "http://github.com/libgit2/TestGitRepository", "./foo", &g_options)) 1: error -1 - Too many redirects", ...)

**Comment by shiftkey** [comment]
:eyes:-ing

**Comment by csware** [comment]
@arrbee @carlosmn Any news on this? AutoCRLF is broken right now, this PR is there for a long time and fixes it...

Btw. These commits are already in TortoiseGIt for half a year, work w/o problems.

**Comment by carlosmn** [comment]
There's still the issue of where the test data comes from AFAICT. This is a situation where we should port git and Git for Windows' tests (if they differ) tests, as what it does at any point is fickle.

**Comment by csware** [comment]
@carlosmn The test data comes from https://download.tortoisegit.org/tmp/crlf-multitest.tar.bz2. I.e. I created a git repository which contains several files with LF-only, CRLF, mixed, UTF-8 and so on. Then I wrote a script which did checkouts + checkins based on this repository on Debian 7 and Git for Wiondows (msys1) and compared all results.
Well, it cannot get worse ;)

**Comment by carlosmn** [comment]
That's some set of files, but how does the script's tests overlap with git's tests? We can't just take a bunch of files and hope they match what git is tested to do.

**Comment by csware** [comment]
So, what I did was to match the source code and also match the objects and files created by git on check-outs and check-ins. Then, I created libgit2 testcases for all the different cases. I.e., libgit2 creates exactly the same files and objects as Git on Windows AND Linux. So the actual output which vanilla Git produces is the gold standard here.

Git for Windows, which I used for creating all the test data, as well as Git on Debian Linux passed the test cases

So, when you download the test files and compare unix.txt and windows.txt you can see the cases where Git for Windows and Git on Linux differ and where not.

**Comment by csware** [comment]
Ok, this is my last try before I close this PR and leave libgit2 with badly broken AutoCrLf and .gitattributes handling... I asked several times for concrete suggestions, but got no answer (when I got feedback from @ethomson I honored all his comments). How got the initial AutoCrLf code get into libgit2? - If AutoCrLf is set on *nix all line Lf endings are also converted to CrLf by vanilla git, but libgit2 doesn't. My tests map the vanilla git behavior 1:1 to libgit2 (verified on code and tracked file level).

The tests correlate with t0025-crlf-auto.sh and t0026-eol-config.sh from vanilla git. Should I exactly map the git tests to libgit2? Right now, Git for Windows and git on Debian pass the test suite and I made libgit2 behave exactly like vanilla git (I tested lots of different combinations of AutoCrLf and .gitattribute settings).

**Comment by ethomson** [comment]
> So the actual output which vanilla Git produces is the gold standard here.

I think that @carlosmn 's point, above, was that you're adding some 2,000 lines of code that we now have to maintain in a manual manner.  We have a _lot_ more unit tests (which is not a bad thing!) but we don't have any way to _produce_ these unit tests.

What I would like to see is a mechanism for _creating these unit tests_.  Because if Git's CRLF behavior changes (and it has, in the past) then we're in the same position we're in now where we're only testing our own implementation of what we think is correct.  And now we have to dig through an additional 2,000 LoC to understand what's going on.

What would be useful here is to:

1. Include a script that could generate these resources
2. Have significantly fewer lines of test code while still testing the same combinations by simply looping through the various combinations and comparing it to the data that the script had produced.
3. Wire up the CI builds so that it runs the script so that we get immediate knowledge that our CRLF handling is broken against some git.git implementation.

**Comment by csware** [comment]
@ethomson The problem here is, that we would need to run git for Windows AND a *nix version of git for producing the test cases.

I suppose that some test cases indeed do overlap. My intension was to check all possible combinations of AutoCrLf, SafeCrLf and .gitattribute settings. I could try to find the overlapping ones and drop them.

**Comment by ethomson** [comment]
That makes sense, and I suppose reduces the likelihood that (3) is actually possible.

My bigger motivation isn't to reduce the overlap.  I'm fine testing overlapping contents, that slows test execution down (which is regrettable, but I'd rather burn CPU than burn a developer's neocortex.

If we have a script that uses git.git to produce this data that you've included and emits files, eg:

    file-autocrlf=true-eol=lf-windows.txt

Then we're able to trivially loop over our options, comparing our output to the file with the name representing the options.  I think this will represent a major reduction in the number of lines of test code and resources, while being easy to generate instead of hard.

Trying to review and understand that we are covering all the possible combinations when we've produced them manually is very complex as - I'm certain - producing them was!  But this is exactly the sort of drudge work that computers excel at, it would be nice to automate this.

**Comment by csware** [comment]
@ethomson I'd also like to automate this using a script, but I'm unsure how to do it. I already spent 4 full workdays on this PR and don't want to see it dropped or just ignored.

Git on *nix and on Windows behave exactly the same for all combinations except for only 3 cases for checkout (for attributes text, text=auto, crlf). Lemme try to reduce the overlapping separate test cases by introducing 4 different functions which get called again and again.

**Comment by csware** [comment]
@ethomson I created two scripts for automatic generation of the test cases (crlf-test-generator.7z in tests folder).

Any ideas on how to integrate it into the tests scripts or just keep it there so that one can it easily re-create the tests?

**Comment by linquize** [comment]
Why do you choose to add a 7z file?

**Comment by csware** [comment]
This is outdated as Git has changed the EOL handling with Git 2.8.

**Pull request closed without merging**